File: longest-increasing-subsequence/longest-increasing-subsequence.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& arr) {
        int n = arr.size();
        int *dp = new int[n]();
        dp[0] = 1;
        int ans = 1;
        for(int i=1;i<n;i++)
        {
            int mx = 0;
            for(int j=0;j<i;j++)
            {
                if(arr[j] < arr[i])
                {
                    mx = max(mx,dp[j]);
                }
            }
            dp[i] = mx  + 1;
            ans = max(ans,dp[i]);
        }
        
        delete[] dp;
        return ans;
    }
};

File: longest-increasing-subsequence/longest-increasing-subsequence.java
class Solution {
    public int lengthOfLIS(int[] arr) {
        int n = arr.length;
        int[] lis = new int[n];
        for(int i=0;i<n;i++) lis[i] = 1;
        
        int i=1;
        while(i<n)
        {
            for(int j=0;j<i;j++)
            {
                if(arr[j] < arr[i] && lis[i] <= lis[j])
                    lis[i] = 1 + lis[j];
            }
            ++i;
        }
        
        int ans = 0;
        for(i=0;i<n;i++) ans = Integer.max(ans,lis[i]);
        
//         String str = "";
        
//         int curr = ans;
        
//         for(i=n-1;i>=0;i--)
//         {
//             if(lis[i] == curr)
//             {
//                 str += String.valueOf(arr[i]);
//                 str += " ";
//                 curr--;
//             }
//         }
        
//         System.out.println(str);
        
        return ans;
    }
}